abouttitle='About'
Aboutinfo_page <- tabPanel(
  theme=shinythemes::shinytheme('flatly'),
  title='About',
  h3("Summary"),
  wellPanel(
    p(paste("We have proposed"),
      HTML(paste0(tags$em("tests of performance sensitivity"))),
      ("that detect whether a polygenic score (PGS) is overfitted to population structure. Our tests work on any PGS, and only require (i) selecting variants that have high GWAS p-value; and (ii) application to a held-out validation cohort not used in training the PGS model. We specifically test for whether the effects of the selected variants are likely causal, under the null hypotheses that the effects are permutation-invariant and that effect directions are random. When these null hypotheses hold, the performance of a PGS with the selected variant effects perturbed or effect directions flipped would not be worse, leading to the PGS exhibiting low performance sensitivity.")),
    p("The core thesis of our work is that PGSs using too many variants are overparameterized, with the overparameterization capturing population structure. Our thesis is justified by the following arguments."),
    tags$li("Theoretically, PGSs can be viewed as linear combinations of PCs, or linear combinations of singular vectors with skewed weights that capture population structure."),
    tags$li(HTML(paste("This implies that choosing completely random variant effects for a PGS leads to 'better than random' performance, if the phenotype also happens to be stratified. We observe this in data from the UK Biobank",HTML(paste0(tags$sup("1"))),"."))),
    tags$li("Empirically, permuting or flipping the signs of some variants — specifically variants with high GWAS p-value — in a PGS does not weaken the performance in a held-out cohort."),
    tags$li("Across phenotypes, the extent to which PGS performance is weakened by the above procedure is driven by stratification of PGS by PCs in the discovery cohort."),
    br(),
    p(paste0("This dashboard accompanies our paper, by allowing the user to analyze our findings (e.g., sensitivity test p-values) at the individual phenotype level. Users may also find the visualizations helpful for gaining intuition about the measures, metrics and tests proposed in our work."))
  ),
  
  h3("Methodology"),
  wellPanel(
    p("Here we provide technical elaborations to ensure our dashboard is sufficiently self-contained. Please refer to the Materials and Methods section of our paper for full details."),
    h4("Tests of Performance Sensitivity and Inflation"),
    p("We introduce three tests (two of which we described above) that can detect the presence of phenotype or PGS bias by population structure. We explain the rationale of each test before briefly describing it. We also provide an infographic below."),
    tags$li(HTML("<b>Performance Inflation Test:</b> If a phenotype is biased by population structure, a PGS constructed by assigning random effects (e.g., having $\\beta_j$'s independently follow the standard normal distribution ) has inflated performance. To detect if the phenotype is at risk of PGS performance inflation, construct many random PGSs (rPGSs, see infographic) on the discovery cohort and evaluate their performance on the phenotype. Next, permute (i.e., randomize) the phenotypes and repeat the performance evaluation. Iterate through the last step many times, and compare the performance of rPGSs on the phenotype against those for the permuted phenotypes.")),
    tags$li(HTML("<b>Performance Sensitivity to Effect Permutations:</b> If a PGS is overfitted to population structure of the discovery cohort, some of the effects of non-significant variants are random rather than likely causal. To detect if the effects of non-significant variants are likely causal, permute the effects of the non-significant variants many times to obtain many permuted PGSs (pPGSs, see infographic). Evaluate the performance of the pPGSs against the phenotype on a separate held-out cohort, relative to the original PGS.")),
    tags$li(HTML("<b>Performance Sensitivity to Effect Sign Flips:</b> A PGS overfitted to population structure of the discovery cohort will also have effect directions of some non-significant variants be random. To detect if the effect directions of non-significant variants are accurate, randomly flip the directionality of effects of the non-significant variants many times to obtain many sign-flipped PGSs (sPGSs, see infographic). Using a separate held-out cohort, evaluate the performance of the sPGSs against the phenotype relative to the original PGS.")),
    br(),
    HTML('<center><img src="outline_graphic.jpg" width="800"></center>'),
    br(),
    p(HTML(paste("Our tests can be viewed as",
            HTML(paste0(tags$em("upstream"))),"and",
            HTML(paste0(tags$em("downstream"))),"diagnostics of the risk of PGS overfitting to population structure. We note that a few other tests have appeared in the literature."))),
    tags$li(HTML("<b>PC Stratification Significance:</b> Appearing in Figure 2A of Sohail and Maier et al.", 
                 HTML(paste0(tags$sup("2"))), ", this downstream diagnostic evaluates if the PGS suffers from population stratification. The authors compute Pearson correlations between the polygenic effect vector and PC loadings, before applying the block-jackknife on variants to decide if the correlation is significantly non-zero.")),
    tags$li(HTML("<b>Performance Inflation of Non-significant Variants:</b> Appearing in Figure 5 of Kerminen et al.",
                 HTML(paste0(tags$sup("3"))), " and Supplementary Figure 6 of Mars et al.",
                 HTML(paste0(tags$sup("4"))), ", this upstream diagnostic evaluates if the GWAS summary statistics used to train a PGS model are insufficiently corrected for population stratification effects. Focusing on geographical stratification in Finland, the authors randomly select approximately independent variants that also have GWAS p-value > 0.5 for a specified target phenotype. They construct random PGSs by simulating Gaussian distributed effects for the selected variants. Differences in the random PGSs between East and West Finnish individuals are subsequently computed, with deviations more extreme than an analytical confidence interval deemed as evidence of uncorrected population stratification.")),
    h4("Association Mapping"),
    p(HTML("Our work investigates quantitative phenotypes only. Following variant quality control protocols as described in Churchhouse et al.", 
      HTML(paste0(tags$sup("5"))),"we perform GWAS on 357,661 unrelated individuals from the UKB, all with self-reported white British ancestry. We considered both raw and inverse rank normalized (IRNT) versions of phenotypes. For the IRNT version, phenotype values were first adjusted for age, sex and other covariates in a phenotype-specific manner as detailed in Fiziev and McRae et al.",
      HTML(paste0(tags$sup("6"))),". GWAS marginal significance of each variant was tested with a linear model that included age, sex, and the first ten genetic principal components as covariates.")),
    h4("PGS Construction"),
    p(HTML("PGSs reported in this dashboard were constructed using Clumping and Thresholding (C&T)",
           HTML(paste0(tags$sup("7"))),", followed by effect size re-weighting via multiple regression on C&T-selected variants in a window-specific manner. For each phenotype, we used $10^{-5}$ and $10^{-8}$ as two p-value thresholds respectively to construct a lenient PGS and a stringent PGS. In our paper, we also analyze 15 mean corpuscular haemoglobin PGSs published in the PGS Catalogue",
           HTML(paste0(tags$sup("8"))),". These include PGSs trained using more sophisticated approaches, such as the elastic net / penalized regression",
           HTML(paste0(tags$sup("9"))),", snpnet",
           HTML(paste0(tags$sup("10"))),", LDpred2",
           HTML(paste0(tags$sup("11"))),", PolyFun-pred",
           HTML(paste0(tags$sup("12"))),", BOLT-LMM",
           HTML(paste0(tags$sup("13")))," and SBayesR",
           HTML(paste0(tags$sup("14"))),"."))
    ),
  h3("Citation"),
  wellPanel(
    p("Aw, A. et al. 'Highly parameterized polygenic scores tend to overfit to population stratification via random effects'. Submitted."),
  ),
  h3("Contact"),
  wellPanel(
    p('For any queries please contact:'),
    p(mailtoR(email = "alan.aw@pennmedicine.upenn.edu",
              text = "Alan Aw"),
      use_mailtoR()),
    p(mailtoR(email = "EliorRahmani@mednet.ucla.edu",
              text = "Elior Rahmani"),
      use_mailtoR()),
    tags$a(href='https://people.eecs.berkeley.edu/~yss/contact.html',
           'Song Lab (PI: Yun Song)'),
    
  ),
  br(),
  h3("References"),
  wellPanel(
    p(span('1.  Bycroft, C. et al. The UK Biobank resource with deep phenotyping and genomic data. '), 
      em('Nature'),strong('562'), span( '203-209 (2018).')),
    p(span('2.	Sohail, M. and Maier, R. et al. Polygenic adaptation on height is overestimated due to uncorrected stratification in genome-wide association studies. '), 
      em('eLife '), strong(' 8'), span(':e39702 (2019).')),
    p(span('3.	Kerminen, S. et al. Geographic Variation and Bias in the Polygenic Scores of Complex Diseases and Traits in Finland. '), 
      em('Am J Hum Genet. '), strong('104'), span(', 1169-1181 (2019).')),
    p(span('4.	Mars, N. et al. The role of polygenic risk and susceptibility genes in breast cancer over the course of life. '), 
      em(' Nat Commun '), strong('11'), span(', 6383 (2020).')),
    p(span('5.	Churchhouse, C. et al. Rapid GWAS of thousands of phenotypes for 337,000 samples in the UK Biobank. '), 
      span('(2017).')),
    p(span('6.	Fiziev, P. and McRae, J. et al. Rare penetrant mutations confer severe risk of common diseases. '), 
      em('Science '), strong('380'), span(', eabo1131 (2023).')),
    p(span('7.	Khera, A. et al. Genome-wide polygenic scores for common diseases identify individuals with risk equivalent to monogenic mutations. '), 
      em('Nat Genet '), strong('50'), span(', 1219-1224 (2018).')),
    p(span('8.	Lambert, S. et al. The Polygenic Score Catalog as an open database for reproducibility and systematic evaluation. '), 
      em('Nat Genet '), strong('53'), span(', 420-425 (2021).')),
    p(span('9.	Tay, J. et al. Elastic Net Regularization Paths for All Generalized Linear Models. '), 
      em('Journal of Statistical Software '), strong('106'), span(', 1-31 (2023).')),
    p(span('10.	Tanigawa, Y. et al. Significant sparse polygenic risk scores across 813 traits in UK Biobank. '), 
      em('PLoS Genetics '), strong('18'), span(': e1010105 (2022).')),
    p(span('11.	Privé, F. et al. LDpred2: better, faster, stronger. '), 
      em('Bioinformatics'), strong('36'), span(', 5424-5431 (2020).')),
    p(span('12.	Weissbrod, O., Kanai, M. and Shi, H. et al. Leveraging fine-mapping and multipopulation training data to improve cross-population polygenic risk scores. '), 
      em('Nat Genet '), strong('54'), span(', 450-458 (2022).')),
    p(span('13.	Loh, P. et al. Efficient Bayesian mixed-model analysis increases association power in large cohorts. '),
      em('Nat Genet '), strong('47'), span(', 284-290 (2015).')),
    p(span('14.	Lloyd-Jones, L. and Zeng, J. et al. Improved polygenic prediction by Bayesian multiple regression on summary statistics. '), 
      em('Nat Commun '), strong('10'), span(', 5086 (2019).'))
  ))
