abouttitle='About'
Aboutinfo_page <- tabPanel(
  theme=shinythemes::shinytheme('flatly'),
  title='About',
  h3("Summary"),
  wellPanel(
    p(paste("We have proposed"),
      HTML(paste0(tags$em("tests of performance sensitivity"))),
      ("that detect whether a polygenic score (PGS) is overfitted to population structure. Our tests work on any PGS, and only require (i) selecting variants that have high GWAS p-value; and (ii) application to a held-out validation cohort not used in training the PGS model. We specifically test for whether the effects of the selected variants are likely causal, under the null hypotheses that the effects are permutation-invariant and that effect directions are random. When these null hypotheses hold, the performance of a PGS with the selected variant effects perturbed or effect directions flipped would not be worse, leading to the PGS exhibiting low performance sensitivity.")),
    p("The core thesis of our work is that PGSs using too many variants are overparameterized, with the overparameterization capturing population structure. Our thesis is justified by the following arguments."),
    tags$li("Theoretically, PGSs can be viewed as linear combinations of PCs, or linear combinations of singular vectors with skewed weights that capture population structure."),
    tags$li(HTML(paste("This implies that choosing completely random variant effects for a PGS leads to 'better than random' performance, if the phenotype also happens to be stratified. We observe this in data from the UK Biobank",HTML(paste0(tags$sup("1"))),"."))),
    tags$li("Empirically, permuting or flipping the signs of some variants — specifically variants with high GWAS p-value — in a PGS does not weaken the performance in a held-out cohort."),
    tags$li("Across phenotypes, the extent to which PGS performance is weakened by the above procedure is driven by stratification of PGS by PCs in the discovery cohort."),
    br(),
    p(paste0("This dashboard accompanies our paper, by allowing the user to analyze our findings (e.g., sensitivity test p-values) at the individual phenotype level. Users may also find the visualizations helpful for gaining intuition about the measures, metrics and tests proposed in our work."))
  ),
  
  h3("Methodology"),
  wellPanel(
    p("Here we provide technical elaborations to ensure our dashboard is sufficiently self-contained. Please refer to the Materials and Methods section of our paper for full details."),
    h4("Tests of Performance Sensitivity and Inflation"),
    p("We introduce three tests (two of which we described above) that can detect the presence of phenotype or PGS bias by population structure. We explain the rationale of each test before briefly describing it. We also provide an infographic below."),
    tags$li(HTML("<b>Performance Inflation Test</b>: If a phenotype is biased by population structure, a PGS constructed by assigning random effects (e.g., having $\\beta_j$'s independently follow the standard normal distribution ) has inflated performance. To detect if the phenotype is at risk of PGS performance inflation, construct many random PGSs (rPGSs, see infographic) on the discovery cohort and evaluate their performance on the phenotype. Next, permute (i.e., randomize) the phenotypes and repeat the performance evaluation. Iterate through the last step many times, and compare the performance of rPGSs on the phenotype against those for the permuted phenotypes.")),
    tags$li(HTML("<b>Performance Sensitivity to Effect Permutations</b>: If a PGS is overfitted to population structure of the discovery cohort, some of the effects of non-significant variants are random rather than likely causal. To detect if the effects of non-significant variants are likely causal, permute the effects of the non-significant variants many times to obtain many permuted PGSs (pPGSs, see infographic). Evaluate the performance of the pPGSs against the phenotype on a separate held-out cohort, relative to the original PGS.")),
    tags$li(HTML("<b>Performance Sensitivity to Effect Sign Flips</b>: A PGS overfitted to population structure of the discovery cohort will also have effect directions of some non-significant variants be random. To detect if the effect directions of non-significant variants are accurate, randomly flip the directionality of effects of the non-significant variants many times to obtain many sign-flipped PGSs (sPGSs, see infographic). Using a separate held-out cohort, evaluate the performance of the sPGSs against the phenotype relative to the original PGS.")),
    HTML('<center><img src="outline_graphic.jpg" width="800"></center>'),
    br(),
    p("Our tests can be viewed as upstream and downstream diagnostics of the risk of PGS overfitting to population structure. We note that a few other tests have appeared in the literature."),
    tags$li(HTML("<b>Performance Inflation of Non-significant Variants</b>")),
    tags$li(HTML("<b>PC Stratification Significance</b>")),
    h4("Association Mapping"),
    p(span(strong("OMIM:"), span("Mendelian genes from the Online Mendelian Inheritance in Man (OMIM) database"),HTML(paste0(tags$sup("4"))),span('.'),
    )),
    p(span(strong("HGMD:"), span("Disease-causing and likely disease-causing genes for human inherited diseases from the human mutation database (HGMD) professional"),HTML(paste0(tags$sup("5"))),span('.'),
    )),
    p(span(strong("EVA-ClinVar:"), span("Clinically relevant genetic variants and diseases from ClinVar"),HTML(paste0(tags$sup("6"))), span(". This evidence is obtained from the Open Target platform and was curated by European Variation Archive (EVA)"),HTML(paste0(tags$sup("7"))), span("ClinVar evidence was filtered on clinical significance: 'likely pathogenic', 'association', 'confers sensitivity', 'drug response', 'protective' and 'pathogenic' and the confidence of the submission was filtered on: 'criteria provided, single submitter'; 'criteria provided, conflicting interpretations'; 'criteria provided, multiple submitters, no conflicts'; 'reviewed by expert panel' and 'practice guideline'."),
    )),
    h4("PGS Construction"),
    p(span(strong("Gene burden:"), span("Exome-based gene burden association statistics from the UK Biobank hosted by Genebass"),HTML(paste0(tags$sup("8"))), span(". We extracted traits labeled ‘ICD first occurrence’ and restricted to missense and pLOF burden sets ("),HTML(paste0(tags$em("P"))),span(" < 4.3 x 10"),HTML(paste0(tags$sup("-7"))),span(")."),
    )),
    p(span(strong("Single-Variant:"), span("Exome-based single variant association statistics from the UK Biobank hosted by Genebass"),HTML(paste0(tags$sup("8"))), span(". We extracted traits labeled ‘ICD first occurrence’ and restricted to missense and pLOF variants ("),HTML(paste0(tags$em("P"))),span(" < 4.3 x 10"),HTML(paste0(tags$sup("-7"))),span(")."),
    )),
    h4("Genome-wide association study phenotypes"),
    
    p(span(strong("eQTL phenotype:"), span("Genes with a GWA phenotype driven by gene expression regulation through shared variants. These phenotypes were obtained from the Pan-UK Biobank"),HTML(paste0(tags$sup("9"))), span("and consisted of 484 ICD-10 traits, 729 phecodes and 39 continuous traits that we had manually mapped to phecodes from the UK Biobank Neale."),
    )),
    p(span(strong("Locus2gene:"), span(" Genes with a GWA phenotype identified by the Locus2Gene machine learning model from the Open Targets Genetics Portal which prioritizes likely causal genes at each GWAS locus by integrating fine-mapped associations with functional genomic features"),HTML(paste0(tags$sup("10"))),span('.'),
    )),
    p(span(strong("pQTL phenotype:"), span(" Genes with a GWA phenotype that are in high linkage disequilibrium (r"),HTML(paste0(tags$sup("2"))),span(">0.80) with at least one "),HTML(paste0(tags$em("cis "))),span("or"), HTML(paste0(tags$em("trans "))) , span("pQTL from supplementary table 12 from the large-scale plasma proteome study conducted by Ferkingstad et al"),HTML(paste0(tags$sup("11"))),span('.'),
    ))),
  h3("Citation"),
  wellPanel(
    p("Aw, A. et al. 'Highly parameterized polygenic scores tend to overfit to population stratification via random effects'. Submitted."),
  ),
  h3("Contact"),
  wellPanel(
    p('For any queries please contact:'),
    p(mailtoR(email = "alan.aw@pennmedicine.upenn.edu",
              text = "Alan Aw"),
      use_mailtoR()),
    p(mailtoR(email = "EliorRahmani@mednet.ucla.edu",
              text = "Elior Rahmani"),
      use_mailtoR()),
    tags$a(href='https://people.eecs.berkeley.edu/~yss/contact.html',
           'Song Lab (PI: Yun Song)'),
    
  ),
  br(),
  h3("References"),
  wellPanel(
    p(span('1.  Bycroft, C. et al. The UK Biobank resource with deep phenotyping and genomic data. '), 
      em('Nature'),strong('562'), span( '203-209 (2018).')),
    p(span('2.	Ochoa, D. et al. Open Targets Platform: supporting systematic drug-target identification and prioritisation. '), em('Nucleic Acids Res '), strong(' 49'), span(', D1302-D1310 (2021).')),
    p(span('3.	Kuhn, M., Letunic, I., Jensen, L.J. & Bork, P. The SIDER database of drugs and side effects. '), em('Nucleic Acids Res '), strong('44'), span(', D1075-D1079 (2016).')),
    p(span('4.	Hamosh, A. et al. Online Mendelian Inheritance in Man (OMIM), a knowledgebase of human genes and genetic disorders. '), em(' Nucleic Acids Res '), strong('30'), span(', 52-5 (2002).')),
    p(span('5.	Stenson, P.D. et al. The Human Gene Mutation Database (HGMD(®)): optimizing its use in a clinical diagnostic or research setting. '), em('Hum Genet '), strong('139'), span(', 1197-1207 (2020).')),
    p(span('6.	Landrum, M.J. et al. ClinVar: improving access to variant interpretations and supporting evidence. '), em('Nucleic Acids Res '), strong('46'), span(', D1062-D1067 (2017).')),
    p(span('7.	Cook, C.E. et al. The European Bioinformatics Institute in 2016: Data growth and integration. '), em('Nucleic Acids Res '), strong('44'), span(', D20-D26 (2015).')),
    p(span('8.	Karczewski, K.J. et al. Systematic single-variant and gene-based association testing of thousands of phenotypes in 394,841 UK Biobank exomes.'), em(' Cell Genomics '), strong('2'), span(', 100168 (2022).')),
    p('9.	Pan-UKB team https://pan.ukbb.broadinstitute.org. (2020).'),
    p(span('10.	Mountjoy, E. et al. An open approach to systematically prioritize causal variants and genes at all published human GWAS trait-associated loci.'), em(' Nat Genet '), strong('53'), span(', 1527-1533 (2021).')),
    p(span('11. Ferkingstad, E. et al. Large-scale integration of the plasma proteome with genetics and disease.'),em(' Nat Genet '),strong('53'), span( '1712-1721 (2021).'))
    
  ))
